plugins/cisco_umbrella_destinations/icon_cisco_umbrella_destinations/util/api.py

```diff
@@ -0,0 +1,191 @@
+
+    # GET list of destinations related to destination list
+    # https://developer.cisco.com/docs/cloud-security/#!get-destinations
+    def get_destinations(self, destination_list_id: int) -> dict:
+        return self._call_api(
+            "GET",
+            f"organizations/{self.org_id}/destinationlists/{destination_list_id}/destinations",
+            None,
+            None,
+            None,
+            action_name="GET list of destinations in destionation list",
+        )
+
+    # POST destinations to existing destination list
+    # https://developer.cisco.com/docs/cloud-security/#!create-destinations
+    def create_destinations(self, destination_list_id: int, data: dict) -> dict:
+        return self._call_api(
+            "POST",
+            f"organizations/{self.org_id}/destinationlists/{destination_list_id}/destinations",
+            None,
+            None,
+            data=data,
+            action_name="Create destinations in existing destination list",
+        )
+
+    # DELETE list of destinations from destination list
+    # https://developer.cisco.com/docs/cloud-security/#!delete-destinations
+    def delete_destinations(self, destination_list_id: int, data) -> dict:
+        return self._call_api(
+            "DELETE",
+            f"organizations/{self.org_id}/destinationlists/{destination_list_id}/destinations/remove",
+            None,
+            None,
+            data=data,
+            action_name="Delete destinations from destination list",
+        )
+
+    def _call_api(
+        self,
+        # method -> GET/POST/etc
+        method: str,
+        # path -> url
+        path: str,
+        # json_data -> Json to send in body
+        json_data: Optional[dict] = None,
+        # Params -> Query (Params) String
+        params: Optional[dict] = None,
+        # data(payload) -> dict in body
+        data: Optional = None,
+        # Action name -> Hard coded written action name for error messages
+        action_name: Optional[str] = None,
+    ) -> dict:
+
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+
+        try:
+            # Prevents json.dumps() on 'None' data
+            if data is None:
+                data_string = None
+            else:
+                data_string = json.dumps(data)
+            response = requests.request(
+                method,
+                self.url + path,
+                json=json_data,
+                params=params,
+                data=data_string,
+                headers=headers,
+                auth=HTTPBasicAuth(self.api_key, self.api_secret),
+            )
+
+            if response.status_code == 401:
+                raise PluginException(
+                    preset=PluginException.Preset.USERNAME_PASSWORD,
+                    cause=f"Invalid request for {action_name}. Invalid API key provided.",
+                    assistance="Verify your API key is correct.",
```

Comment:
`cause` and `assistance` arguments are ignored if a `preset` is provided - I would just remove those and use just the `preset` here (and the below scenarios). No need to provide the `action_name` since that will come back automatically in the logs associated with any failures. I like your thinking around this, though!